vxlan/vxlan.go

```diff
@@ -0,0 +1,277 @@
+// Package vxlan provides the capabilities to create a VXLAN overlay network
+package vxlan
+
+import (
+	"net"
+	"sync"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/pkg/errors"
+	"github.com/rancher/go-rancher-metadata/metadata"
+	"github.com/vishvananda/netlink"
+)
+
+const (
+	changeCheckInterval = 5
+	metadataURL         = "http://rancher-metadata/2016-07-29"
+	vxlanMACRange       = "0E:00:00:00:00:00"
+	vxlanPort           = 4789
+)
+
+const (
+	DefaultBridgeName             = "docker0"
+	DefaultVxlanMTU               = 1500
+	DefaultDisableChecksumOffload = false
+	DefaultPhysicalInterfaceName  = "eth0"
+	DefaultVxlanVNI               = 1042
+	DefaultVxlanInterfaceName     = "vtep1042"
+)
+
+// Vxlan is used to store the VXLAN overlay information
+type Vxlan struct {
+	mu sync.Mutex
+	m  metadata.Client
+	v  *vxlanIntfInfo
+	h  metadata.Host
+
+	BridgeName             string
+	VxlanMTU               int
+	PhysicalInterfaceName  string
+	DisableChecksumOffload bool
+	VxlanVNI               int
+	VxlanInterfaceName     string
+}
+
+// NewVxlan is used to create a new VXLAN Overlay network
+func NewVxlan() (*Vxlan, error) {
+	m, err := metadata.NewClientAndWait(metadataURL)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to create metadata client")
+	}
+
+	selfHost, err := m.GetSelfHost()
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to get self host from metadata")
+	}
+
+	o := &Vxlan{
+		m: m,
+		h: selfHost,
```

Comment:
Why are you storing the host information here? Please don't. This can keep changing later. Always fetch it when it's needed.